### PR TITLE
Issue #1938: If in customer interface, use UserID of initiating user

### PR DESCRIPTION
**Pull request by stefanhaerter**
Closing #1938

### CODE REVIEW

**Review by svenoe — COMMENTED** [review]
In my opinion this is not the correct way to fix things. Just because CustomerUserB prints the ticket of CustomerUserA, still the CustomerData of the ticket should be the data of CustomerA. If this is used nowhere else, but just in the "printed by" field, then there would be no wrong effects, but I would assume this is not the case. I think the "printed by" has to be solved more locally (without having have looked at it, though).

File: Kernel/Output/PDF/Ticket.pm [hunk 1]
```diff
@@ -173,11 +173,8 @@ sub GeneratePDF {
...
     my $CustomerUserObject = $Kernel::OM->Get('Kernel::System::CustomerUser');
     my %CustomerData;
-    if ( $Interface{Customer} ) {
-        %CustomerData = $CustomerUserObject->CustomerUserDataGet(
-            User => $Param{UserID},
-        );
-    }
+    my %PrintingCustomerData;
+
     elsif ( $Ticket{CustomerUserID} ) {
```
Comment (bschmalhofer) [inline Kernel/Output/PDF/Ticket.pm]: Shouldn't this be an 'if' instead of an 'elsif' ?

Comment (stefanhaerter) [reply Kernel/Output/PDF/Ticket.pm]: Yes, it should - strange, I thought I pushed the correction.

File: Kernel/Output/PDF/Ticket.pm [hunk 2]
```diff
@@ -184,6 +186,17 @@ sub GeneratePDF {
         );
     }
 
+    if ( $Interface{Customer} ) {
+        %PrintingCustomerData = $CustomerUserObject->CustomerUserDataGet(
+            User => $Param{UserID},
+        );
+    }
+
+    if ( !defined %PrintingCustomerData )
```
Comment (bschmalhofer) [inline Kernel/Output/PDF/Ticket.pm]: Checking the definedness of hashes and arrays is not really supported. The documentation says:
            Use of "defined" on aggregates (hashes and arrays) is no longer
            supported. It used to report whether memory for that aggregate
            had ever been allocated. You should instead use a simple test
            for size:

                if (@an_array) { print "has array elements\n" }
                if (%a_hash)   { print "has hash members\n"   }

File: Kernel/Output/PDF/Ticket.pm [hunk 3]
```diff
@@ -184,6 +186,17 @@ sub GeneratePDF {
         );
     }
 
+    if ( $Interface{Customer} ) {
+        %PrintingCustomerData = $CustomerUserObject->CustomerUserDataGet(
+            User => $Param{UserID},
+        );
+    }
+
+    if ( !%PrintingCustomerData )
```
Comment (bschmalhofer) [inline Kernel/Output/PDF/Ticket.pm]: Looks better. Did you run the CodePolicy on that code? I think that Perl::Critic would put the opening brace at the end of line 195.

Comment (stefanhaerter) [reply Kernel/Output/PDF/Ticket.pm]: CodePolicy didn't change anything - made this change by myself.

File: Kernel/Output/PDF/Ticket.pm [hunk 4]
```diff
@@ -258,9 +258,7 @@ sub GeneratePDF {
         'DateFormat',
     );
     if ( $Interface{Agent} ) {
-        $PrintedBy .= ' ' . $TicketInfo{User}{UserFullname} . ' ('
-            . $TicketInfo{User}{UserEmail} . ')'
-            . ', ' . $Time;
+        $PrintedBy .= " $TicketInfo{User}{UserFullname} ($TicketInfo{User}{UserEmail}), $Time";
```
Comment (bschmalhofer) [inline Kernel/Output/PDF/Ticket.pm:261]: I would always use the long form with nested hashes, but I'm not sure whether that convention has been agreed upon:
`$TicketInfo{User}{UserFullname}  => $TicketInfo{User}->{UserFullname} `

Comment (stefanhaerter) [reply Kernel/Output/PDF/Ticket.pm:261]: As discussed, I would leave this as it is, regarding the pre-existing code style.

**Review by bschmalhofer — APPROVED** [review]
No objections from my side.

**Pull request merged**